File: architecture-skills/design-patterns/examples/decorator/decorator.go
// Decorator pattern in Go.
//
// Each add-on embeds the wrapped Beverage interface: methods that are
// not overridden would be promoted automatically, so a decorator only
// writes the behavior it changes. http.Handler middleware uses the
// same wrapping technique.

package main

import "fmt"

type Beverage interface {
	Description() string
	Cost() float64
}

type Espresso struct{}

func (Espresso) Description() string { return "espresso" }
func (Espresso) Cost() float64       { return 2.00 }

type Milk struct{ Beverage }

func (m Milk) Description() string { return m.Beverage.Description() + " + milk" }
func (m Milk) Cost() float64       { return m.Beverage.Cost() + 0.30 }

type Mocha struct{ Beverage }

func (m Mocha) Description() string { return m.Beverage.Description() + " + mocha" }
func (m Mocha) Cost() float64       { return m.Beverage.Cost() + 0.50 }

type WhippedCream struct{ Beverage }

func (w WhippedCream) Description() string { return w.Beverage.Description() + " + whipped cream" }
func (w WhippedCream) Cost() float64       { return w.Beverage.Cost() + 0.40 }

func main() {
	plain := Espresso{}
	fancy := WhippedCream{Mocha{Milk{Espresso{}}}}
	for _, drink := range []Beverage{plain, fancy} {
		fmt.Printf("%s: $%.2f\n", drink.Description(), drink.Cost())
	}
}

File: architecture-skills/design-patterns/examples/decorator/decorator.c
/* Decorator pattern in C.
 *
 * Polymorphism is emulated with structs and function pointers: every
 * beverage exposes description/cost through the same struct layout,
 * and a decorator holds a pointer to the beverage it wraps.
 */

#include <stdio.h>

typedef struct Beverage {
    void (*description)(const struct Beverage *self, char *out, size_t cap);
    double (*cost)(const struct Beverage *self);
    const struct Beverage *inner; /* NULL for base beverages */
    const char *label;
    double price; /* base price, or surcharge for add-ons */
} Beverage;

static void base_description(const Beverage *self, char *out, size_t cap) {
    snprintf(out, cap, "%s", self->label);
}

static double base_cost(const Beverage *self) {
    return self->price;
}

static void addon_description(const Beverage *self, char *out, size_t cap) {
    char inner[128];
    self->inner->description(self->inner, inner, sizeof(inner));
    snprintf(out, cap, "%s + %s", inner, self->label);
}

static double addon_cost(const Beverage *self) {
    return self->inner->cost(self->inner) + self->price;
}

static Beverage espresso(void) {
    Beverage b = {base_description, base_cost, NULL, "espresso", 2.00};
    return b;
}

static Beverage add_on(const Beverage *inner, const char *label, double surcharge) {
    Beverage b = {addon_description, addon_cost, inner, label, surcharge};
    return b;
}

static void print_drink(const Beverage *drink) {
    char text[128];
    drink->description(drink, text, sizeof(text));
    printf("%s: $%.2f\n", text, drink->cost(drink));
}

int main(void) {
    Beverage plain = espresso();

    Beverage base = espresso();
    Beverage with_milk = add_on(&base, "milk", 0.30);
    Beverage with_mocha = add_on(&with_milk, "mocha", 0.50);
    Beverage fancy = add_on(&with_mocha, "whipped cream", 0.40);

    print_drink(&plain);
    print_drink(&fancy);
    return 0;
}
